sdk/data/azcosmos/cosmos_client_retry_policy.go

```diff
@@ -0,0 +1,137 @@
+				if !p.attemptRetryOnServiceUnavailable(req, o.isWriteOperation) {
+					p.retryCount = 0
+					p.sessionRetryCount = 0
+					break
+				}
+			}
+			fmt.Println("bout to retry this")
+			response, err = req.Next()
+			fmt.Println("should have retried")
+		} else {
+			fmt.Println("supposed to break this")
+			p.retryCount = 0
+			p.sessionRetryCount = 0
+			break
+		}
+	}
+
+	return response, err
+}
+
+func (p *clientRetryPolicy) shouldRetryStatus(status int, subStatus string) (shouldRetry bool) {
+	if (status == 403 && (subStatus == "3" || subStatus == "1008")) ||
+		(status == 404 && subStatus == "1002") ||
+		(status == 403 && subStatus == "3") ||
+		(status == 503) || status == 409 { // remove 409 for real policy
+		return true
+	}
+	return false
+}
+
+func (p *clientRetryPolicy) attemptRetryOnEndpointFailure(req *policy.Request, isWriteOperation bool) bool {
+	if (p.retryCount > maxRetryCount) || !p.gem.locationCache.enableEndpointDiscovery {
+		return false
+	}
+	if isWriteOperation {
+		p.gem.MarkEndpointUnavailableForWrite(*req.Raw().URL)
+	} else {
+		p.gem.MarkEndpointUnavailableForRead(*req.Raw().URL)
+	}
+	p.gem.Update(req.Raw().Context())
+
+	resolvedEndpoint := p.gem.ResolveServiceEndpoint(p.retryCount, isWriteOperation)
+	req.Raw().Host = resolvedEndpoint.Host
+	req.Raw().URL.Host = resolvedEndpoint.Host
+	p.retryCount += 1
+	time.Sleep(defaultBackoff * time.Second)
+	return true
+}
+
+func (p *clientRetryPolicy) attemptRetryOnSessionUnavailable(req *policy.Request, isWriteOperation bool) bool {
+	if p.gem.CanUseMultipleWriteLocations() {
+		endpoints := []string{}
+		if isWriteOperation {
+			endpoints = p.gem.locationCache.locationInfo.availWriteLocations
+		} else {
+			endpoints = p.gem.locationCache.locationInfo.availReadLocations
+		}
+		if p.sessionRetryCount >= len(endpoints) {
+			return false
+		}
+		resolvedEndpoint := p.gem.ResolveServiceEndpoint(p.retryCount, isWriteOperation)
+		req.Raw().Host = resolvedEndpoint.Host
+		req.Raw().URL.Host = resolvedEndpoint.Host
+	} else {
+		if p.sessionRetryCount > 0 {
+			return false
+		}
+	}
+	p.sessionRetryCount += 1
+	return true
+}
+
+func (p *clientRetryPolicy) attemptRetryOnServiceUnavailable(req *policy.Request, isWriteOperation bool) bool {
+	//On HTTP 503 response, if it's a read request and preferredRegions > 1,
+	//retry on the next preferredRegion. If it's a write request and account is multi master
+	//and preferredRegions > 1, retry on the next preferredRegion.
+	if !p.gem.locationCache.enableEndpointDiscovery || p.preferredLocationIndex >= len(p.gem.preferredLocations) {
+		return false
+	}
+	if isWriteOperation {
+		if p.gem.CanUseMultipleWriteLocations() {
+			locationalEndpoint := p.gem.GetPreferredLocationEndpoint(p.preferredLocationIndex, *req.Raw().URL)
+			req.Raw().URL = &locationalEndpoint
+		} else {
+			return false
+		}
+	} else {
+		locationalEndpoint := p.gem.GetPreferredLocationEndpoint(p.preferredLocationIndex, *req.Raw().URL)
```

Comment:
Why can't we use `p.gem.ResolveServiceEndpoint(p.retryCount, isWriteOperation)`?

Reply:
we can, but it is not the same - if a user doesn't supply preferred regions then it will not be populated. 

I think this might have been a gap on my end though, from reading the description on 503 retries I somehow figured we would be needing to use the actual user-supplied preferred regions, but I see in the different SDKs we just use a flag and then use resolveServiceEndpoint

Reply:
ResolveServiceEndpoint uses the location cache, location cache Write/Read Endpoints are populated taking into account the preferred regions:

https://github.com/Azure/azure-sdk-for-go/blob/10dd3c522da479beb9dc1330c6863b7cebcc6e18/sdk/data/azcosmos/cosmos_location_cache.go#L97-L98